contracts/wallet/WalletFactory.sol

```diff
@@ -194,10 +170,28 @@ contract WalletFactory is Owned, Managed {
         view
         returns (address _wallet)
     {
-        bytes32 newsalt = keccak256(abi.encodePacked(_salt, _owner, _modules));
-        bytes memory code = abi.encodePacked(type(Proxy).creationCode, uint256(walletImplementation));
-        bytes32 hash = keccak256(abi.encodePacked(bytes1(0xff), address(this), newsalt, keccak256(code)));
-        _wallet = address(uint160(uint256(hash)));
+        _wallet = _getAddressForCounterfactualWallet(_owner, _modules, address(0), _salt);
+    }
+
+    /**
+     * @dev Gets the address of a counterfactual wallet with a first default guardian.
+     * @param _owner The account address.
+     * @param _modules The list of modules.
+     * @param _guardian The guardian address.
+     * @param _salt The salt.
+     * @return the address that the wallet will have when created using CREATE2 and the same input parameters.
+     */
+    function getAddressForCounterfactualWalletWithGuardian(
+        address _owner,
+        address[] calldata _modules,
+        address _guardian,
```

Comment:
We should validate the guardian address is non empty here as it's not done further down the call chain

Reply:
I don't agree. If the guardian is empty (address(0)) it will be ignored both by this method and when the wallet is created by calling `createCounterfactualWalletWithGuardian()` with an empty guardian. So the address will match the created wallet as expected.

Reply:
Ok I actually see your point. It's misleading to call `createCounterfactualWalletWithGuardian()` and have no guardian added. I've fixed it.